### ISSUE TITLE
Text measure on server-side-deployment gets inaccurate measurements for SVG output

### ISSUE THREAD

**Issue by dankrusi** [vega/vega#2940]
There is a bug in vega-scenagraph text util: when getting SVG output on server-side-deployment (ie NodeJS), the measurement is slightly off. This is most noticable in small fonts that push a layouts boundary, such as a left axis with right-aligned labels.

### Expected
![image](https://user-images.githubusercontent.com/190233/96873470-44077280-1475-11eb-97a3-0d4140c2bbd0.png)

### Results
![image](https://user-images.githubusercontent.com/190233/96877025-69967b00-1479-11eb-85e3-b94a7798082d.png)

## Reproduce

- Take any example from https://vega.github.io/vega/examples/ with a left axis.
- Remove the chart global padding (this hides the bug)
- Increase the numbers so that the axis labels have greater compounded error in measurement
- Set the font to a known machine font, like "Times New Roman"
- Generate a SVG on NodeJS
- Open the SVG in a standard SVG viewer, such as a chromium-based browser or Adobe Illustrator

or

- Run ````node vega-bug-font-measure-nodejs-svg-out_v1.js```

Note: make sure you have the node-canvas package installed (```npm install canvas```)

## Affects

To my knowledge all recent vega versions.

## Solution

It seems to be an issue with the way vega initializes its 1x1 canvas for text measuring: in effect vega always uses a node-canvas raster type (ie using the constructor ```NodeCanvas.Canvas(w, h, null);```, when it should be using a SVG type (```NodeCanvas.Canvas(w, h, 'svg');```) when outputting to SVG. This is because node-canvas/Cairo/Pango will return slightly different measurement values:

```
$ node node-canvas-cairo-font-measurement-raster.js
{
  width: 44, // via NodeCanvas.Canvas(w, h)
  actualBoundingBoxLeft: 0,
  actualBoundingBoxRight: 44,
  actualBoundingBoxAscent: 7,
  actualBoundingBoxDescent: 1,
  emHeightAscent: 10,
  emHeightDescent: 2,
  alphabeticBaseline: -0
}
$ node node-canvas-cairo-font-measurement-svg.js
{
  width: 40.5, // via NodeCanvas.Canvas(w, h, 'svg')
  actualBoundingBoxLeft: 0,
  actualBoundingBoxRight: 40.5,
  actualBoundingBoxAscent: 6.08203125,
  actualBoundingBoxDescent: 1.4990234375,
  emHeightAscent: 10,
  emHeightDescent: 2,
  alphabeticBaseline: -0
}
```

Setting the measuring canvas constructor type to SVG produces more accurate font metrics and the axis labels are not cuttoff. One would need to have different types of canvases to measure on, depending on what the current output is.

### Additional Problem

After attempting to create a fix and some further digging, it seems that node-canvas/Cario/Pango lib locks in the measureText backing type to the first canvas type instanced. This means that in the following:

```
canvas1 = new canvas with raster type
canvas1 measureText // outputs value for raster: 44
canvas2 = new canvas with svg type
canvas2 measureText // outputs value for raster: 44 (but we expect the svg results)
```

## Infos

Vega v5.17.0
NodeJS v12.19.0
node-canvas v2.6.1

Build Environment: Windows 10.0.18363
Test Environment: Windows 10.0.18363

## Spec

```json
{
  "$schema": "https://vega.github.io/schema/vega/v5.json",
  "description": "A basic stacked bar chart example.",
  "width": 500,
  "height": 200,
  "padding": 0,
  "data": [
    {
      "name": "table",
      "values": [
        {"x": 0, "y": 28000000, "c": 0}, {"x": 0, "y": 55000000, "c": 1},
        {"x": 1, "y": 43000000, "c": 0}, {"x": 1, "y": 91000000, "c": 1},
        {"x": 2, "y": 81000000, "c": 0}, {"x": 2, "y": 53000000, "c": 1},
        {"x": 3, "y": 19000000, "c": 0}, {"x": 3, "y": 87000000, "c": 1},
        {"x": 4, "y": 52000000, "c": 0}, {"x": 4, "y": 48000000, "c": 1},
        {"x": 5, "y": 24000000, "c": 0}, {"x": 5, "y": 49000000, "c": 1},
        {"x": 6, "y": 87000000, "c": 0}, {"x": 6, "y": 66000000, "c": 1},
        {"x": 7, "y": 17000000, "c": 0}, {"x": 7, "y": 27000000, "c": 1},
        {"x": 8, "y": 68000000, "c": 0}, {"x": 8, "y": 16000000, "c": 1},
        {"x": 9, "y": 49000000, "c": 0}, {"x": 9, "y": 15000000, "c": 1}
      ],
      "transform": [
        {
          "type": "stack",
          "groupby": ["x"],
          "sort": {"field": "c"},
          "field": "y"
        }
      ]
    }
  ],

  "scales": [
    {
      "name": "x",
      "type": "band",
      "range": "width",
      "domain": {"data": "table", "field": "x"}
    },
    {
      "name": "y",
      "type": "linear",
      "range": "height",
      "nice": true, "zero": true,
      "domain": {"data": "table", "field": "y1"}
    },
    {
      "name": "color",
      "type": "ordinal",
      "range": "category",
      "domain": {"data": "table", "field": "c"}
    }
  ],

  "axes": [
    {"orient": "bottom", "scale": "x", "zindex": 1},
    {
      "orient": 
      "left", 
      "scale": "y", 
      "zindex": 1,
      "labelFont": "'Times New Roman'",
      "labelFontSize": 15
    }
  ],

  "marks": [
    {
      "type": "rect",
      "from": {"data": "table"},
      "encode": {
        "enter": {
          "x": {"scale": "x", "field": "x"},
          "width": {"scale": "x", "band": 1, "offset": -1},
          "y": {"scale": "y", "field": "y0"},
          "y2": {"scale": "y", "field": "y1"},
          "fill": {"scale": "color", "field": "c"}
        },
        "update": {
          "fillOpacity": {"value": 1}
        },
        "hover": {
          "fillOpacity": {"value": 0.5}
        }
      }
    }
  ]
}
```

## Example Files (zip) 

[vega-bug-font-measure-nodejs-svg-out_v1.zip](https://github.com/vega/vega/files/5423021/vega-bug-font-measure-nodejs-svg-out_v1.zip)

## Related

https://github.com/vega/vega/issues/2396

**Comment by domoritz** [comment]
Thanks for the bug report. Are you suggesting that we should pass the type to https://github.com/vega/vega/blob/master/packages/vega-scenegraph/src/util/canvas/context.js#L3? Would that be sufficient?

**Comment by dankrusi** [comment]
Yes exactly. If you are exporting to SVG you need to do text metrics on a node-canvas type of svg. Then you have better results with the SVG.

I am not so familiar with the codebase of vega, so I don't have the concrete solution. But my guess is that the text measure should be refactored to the Renderer, and the renderer should you a different static/cached canvas depending on the output.

**Comment by kachkaev** [comment]
I’m facing this issue when generating an svg vega chat in Node.js. Interestingly, font metrics can be fixed just by installing [node-canvas](https://github.com/Automattic/node-canvas). This creates another problem though because my project also depends on `sharp`: #1386.

I understand that getting font metrics to work in Node.js is a pretty hard problem. However I wonder if there could be any way to disable space calculations based on labels? If I could tell vega to not look at text width of the title as well as the x axis labels, the unwanted padding I get in the svg would disappear.

**Comment by jheer** [comment]
Yes, Vega uses Canvas (either in the browser, or via node-canvas) to calculate text metrics, regardless of the output renderer. So in general, one should include node-canvas if one needs accurate font metrics for server-side rendering.

Vega uses a text metrics object that is exported by vega-scenegraph and then vega itself: https://github.com/vega/vega/blob/master/packages/vega-scenegraph/src/util/text.js#L7

One approach could be to try overwriting the `estimateWidth` method (e.g., to return zero) in your own application. This is a bit of a hack but might suffice in a specialized case.

Or, one could consider a more thorough update to support more customized estimation. I'm open to PRs for further customizing / disabling text metrics. 

p.s. @kachkaev I can't follow the issue link you included as it points to Vega issue #1386, which is about Vega logos... not a node-canvas related issue.

**Comment by RedShift1** [comment]
Any workaround without having to install a canvas renderer?
